apps/prairielearn/elements/pl-excalidraw/pl-excalidraw.py

```diff
@@ -0,0 +1,133 @@
+import base64
+import json
+from pathlib import Path
+
+import chevron
+import lxml.html
+import prairielearn as pl
+from lxml.html import HtmlElement
+from typing_extensions import assert_never
+
+ATTR_ANSWER_NAME = "answers-name"
+ATTR_WIDTH = "width"
+ATTR_HEIGHT = "height"
+ATTR_SOURCE_FILE_NAME = "source-file-name"
+ATTR_SOURCE_DIRECTORY = "directory"
+
+
+SOURCE_DIRECTORY_MAP = {
+    "serverFilesCourse": "server_files_course_path",
+    "clientFilesCourse": "client_files_course_path",
+    "clientFilesQuestion": "client_files_question_path",
+    "courseExtensions": "course_extensions_path",
+    ".": "question_path",
+}
+
+
+def prepare(element_html: str, data: pl.QuestionData) -> None:
+    element = lxml.html.fragment_fromstring(element_html)
+    required_attrs = [ATTR_ANSWER_NAME]
+    optional_attrs = [
+        ATTR_WIDTH,
+        ATTR_HEIGHT,
+        ATTR_SOURCE_FILE_NAME,
+        ATTR_SOURCE_DIRECTORY,
+    ]
+    pl.check_attribs(element, required_attrs, optional_attrs)
+
+    name = pl.get_string_attrib(element, ATTR_ANSWER_NAME)
+    pl.check_answers_names(data, name)
+
+    source_dir = pl.get_string_attrib(element, ATTR_SOURCE_DIRECTORY, ".")
+    if source_dir not in SOURCE_DIRECTORY_MAP:
+        raise RuntimeError(
+            f"{source_dir=} has to be one of {list(SOURCE_DIRECTORY_MAP.keys())}"
+        )
+
+
+def load_file_content(element: HtmlElement, data: pl.QuestionData) -> str:
+    file_dir = SOURCE_DIRECTORY_MAP[
+        pl.get_string_attrib(element, ATTR_SOURCE_DIRECTORY, ".")
+    ]
+    file = Path(data["options"][file_dir]) / pl.get_string_attrib(
+        element, ATTR_SOURCE_FILE_NAME
+    )
+    if not file.exists():
+        raise RuntimeError(f"Unknown file path: {file}")
+    return file.read_text(encoding="utf-8")
+
+
+def render(element_html: str, data: pl.QuestionData) -> str:
+    element = lxml.html.fragment_fromstring(element_html)
+    drawing_name = pl.get_string_attrib(element, ATTR_ANSWER_NAME)
+    initial_content: str = ""
+
+    match data["panel"]:
+        case "answer":
+            # Answer must have a file attribute
+            if not pl.has_attrib(element, ATTR_SOURCE_FILE_NAME):
+                raise RuntimeError(
+                    f"Answer drawing '{drawing_name}' does not have a `file` argument"
+                )
+            initial_content = load_file_content(element, data)
```

Comment:
I'm thinking a bit about the ergonomics of this element. From what I can see, if you provide a `source-file-name` attribute and unconditionally render this element without e.g. `pl-hide-in-panel`, the "starter" drawing will be presented as the answer. This deviates from host most other PL elements work. At the same time, this element is already a bit unique in that it's used to both render "things" (prompt diagrams, answers, etc.) from files _and_ accept student input. If one wants to use it strictly for display, one still has to specify an `answers-name`, which feels very strange to me.

Maybe this is all fine, but I want to make sure we've carefully thought through this element's API. One route we could take (for discussion) is to support an attribute like `type="input"`/`type="display"`. The former would make `answers-name` required and would render nothing in the `answer` panel; the latter would require `source-file-name` and `directory` required and would render in all panels. This would have a nice benefit that a "display-only" drawing that's rendered in the question panel wouldn't end up submitting data.

I'll raise this for discussion with the core team on Slack and get back to you soon.

Reply:
See https://prairielearn.readthedocs.io/en/latest/pl-drawing/

Reply:
Specifically:

- Add a `gradable` attribute, default to `true`. Match the behavior discovered in `<pl-drawing>`.
- If `gradable` is `false`, don't render in the answer panel; a student submission should never appear in the answer panel.

How does this sound?

Reply:
I'm writing this down to confirm we're on the same page :)

Listing out the variables in our matrix, we have the presence of source-file-name, if the render is fresh or a submission exists, gradable, and the panel type.

RO/RW - ReadOnly/ReadWrite
ANS - Require `answers-name` attribute

**Question**

- Gradable (RW) (ANS)
  - Fresh
    - Source: Render the source
    - NoSource: Empty diagram
  - Submission: Render the submission
- NonGradable (RO)
  - Fresh
    - Source: Render the source
    - NoSource: Raise an error
  - Submission: Invalid state

**Answer** (RO)

- Gradable: Hide (ANS)
- NonGradable
  - Source
    - Fresh: Render the source
    - Submission: Invalid state
  - NoSource: Raise an error

**Submission** (RO)

- Gradable (ANS)
  - Fresh: Hide
  - Submission: Render the submission
- NonGradable: Hide

Providing `answers-name` depends on if a widget is gradable, so the documentation is slightly more verbose.

Does this model look okay to you?

Reply:
A few notes:

* "Question -> RO -> Submission" shouldn't explicitly be considered an invalid state, though I guess this depends on exactly how this state is defined. In my mind this is just "the question panel is rendered again after a student submits", in which case we just render whatever the source file is.
* "Submission -> Gradable -> Fresh" is actually a state we can't end up in (the submission panel is never rendered before there's a submission). This almost certainly shouldn't impact the code you write.
* "Submission -> NonGradable" shouldn't explicitly result in a "hide" in the code you write. It's valid for an instructor to render "display-only" elements in any panel, and this is what happens by default if an element isn't nested in `<pl-question-panel>`/etc.

Reply:
Updated the combinations above with your feedback. `invalid_state` translates to `raise unreachable` in code for the sake of exhaustiveness checking.